p2p/utils.go

```diff
@@ -0,0 +1,41 @@
+package p2p
+
+import "sync"
+
+type ConnectCallbacks struct {
+	cbs []ConnectCallback
+	mu  sync.Mutex
+}
+
+func (a *ConnectCallbacks) Add(cb ConnectCallback) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	a.cbs = append(a.cbs, cb)
+}
+
+func (a *ConnectCallbacks) GetAll() []ConnectCallback {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	out := make([]ConnectCallback, len(a.cbs))
```

Comment:
If you changed it to RWMutexm, then here we use RLock and RUnlock.

Reply:
changed